### DISCUSSION TITLE
Markdown Support

### DISCUSSION THREAD

**Discussion by xin2017338** [biomejs/biome#3816] [RFC]
markdown is also already in prettify and is what you need.

So our markdown support is also important, using this pr to support basic markdown

For now, I'm going to use this ebnf syntax and see if I can match it with prettify later.

https://www.mediawiki.org/wiki/Markup_spec/EBNF

Why not an existing library?

Because converting from an existing ast to biome's cst and ast is tedious and not always possible, just like the css implementation, we can achieve a more fine-grained improvement by doing it ourselves.

This pr only implements the ast part of markdown, we will add support for format/lint when the ast part is stable

#3718 

Reference Links

https://www.mediawiki.org/wiki/Markup_spec/EBNF

https://github.github.com/gfm/

https://github.com/syntax-tree/mdast

https://commonmark.org/

Progress tracking

- [x] project struct init #3775 
  - [x] grammer support
  - [ ] support commonmark grammer defined
   
- grammer implementation

  - block 
    -  leaf block
      - [ ] [4.1Thematic breaks](https://spec.commonmark.org/0.31.2/#thematic-breaks) https://github.com/biomejs/biome/pull/3982
      - [ ] [4.2ATX headings](https://spec.commonmark.org/0.31.2/#atx-headings)
      - [ ] [4.3Setext headings](https://spec.commonmark.org/0.31.2/#setext-headings)
      - [ ] [4.4Indented code blocks](https://spec.commonmark.org/0.31.2/#indented-code-blocks)
      - [ ] [4.5Fenced code blocks](https://spec.commonmark.org/0.31.2/#fenced-code-blocks)
      - [ ] [4.6HTML blocks](https://spec.commonmark.org/0.31.2/#html-blocks)
      - [ ] [4.7Link reference definitions](https://spec.commonmark.org/0.31.2/#link-reference-definitions)
      - [ ] [4.8Paragraphs](https://spec.commonmark.org/0.31.2/#paragraphs)
      - [ ] [4.9Blank lines](https://spec.commonmark.org/0.31.2/#blank-lines)
    -  container bloock
      - [ ] [5.1Block quotes](https://spec.commonmark.org/0.31.2/#block-quotes)
      - [ ] [5.2List items](https://spec.commonmark.org/0.31.2/#list-items)
      - [ ] [5.3Lists](https://spec.commonmark.org/0.31.2/#lists)
  
  - inline block
    - [ ] [6.1Code spans](https://spec.commonmark.org/0.31.2/#code-spans)
    - [ ] [6.2Emphasis and strong emphasis](https://spec.commonmark.org/0.31.2/#emphasis-and-strong-emphasis)
    - [ ] [6.3Links](https://spec.commonmark.org/0.31.2/#links)
    - [ ] [6.4Images](https://spec.commonmark.org/0.31.2/#images)
    - [ ] [6.5Autolinks](https://spec.commonmark.org/0.31.2/#autolinks)
    - [ ] [6.6Raw HTML](https://spec.commonmark.org/0.31.2/#raw-html)
    - [ ] [6.7Hard line breaks](https://spec.commonmark.org/0.31.2/#hard-line-breaks)
    - [ ] [6.8Soft line breaks](https://spec.commonmark.org/0.31.2/#soft-line-breaks)
    - [ ] [6.9Textual content](https://spec.commonmark.org/0.31.2/#textual-content)
    - [ ] [Appendix: A parsing strategy](https://spec.commonmark.org/0.31.2/#appendix-a-parsing-strategy)

- extension support 
  - [ ] gfm support

**Comment by Conaclos** [comment]
Are we aiming for compliance to the [CommonMark spec](https://commonmark.org/)?

**Reply by xin2017338** [reply]
The ultimate goal is to finish making commonmark spec at https://github.github.com/gfm/

**Reply by AugustinMauroy** [reply]
GFM must be a option it's shouldn't be the default option

**Comment by AugustinMauroy** [comment]
I was wondering if it was possible to support MDX easily. Because biome already supports JSX.

**Reply by dyc3** [reply]
Probably doable, but likely requires #3334

**Reply by ematipico** [reply]
Supporting MDX to what end? 

MDX is a completely different language from markdown, and it has a lot of different parsing rules. "Easily" might not be the right word 😅

**Reply by AugustinMauroy** [reply]
Okay good to know. 👍

**Reply by AugustinMauroy** [reply]
IDK what is the politic of biome on create. Bu there are a mdx parser https://github.com/wooorm/mdxjs-rs so is it possible to use on biome ?

**Reply by ematipico** [reply]
Technically speaking, the only way to support other parsers inside Biome, is to map the other parser nodes (AST/CST) to Biome CST.

I'm not sure what you mean by "politic". We simply haven't tried this approach and we don't have the bandwidth to explore it, so you're more than welcome to explore this solution and see if it's viable

**Reply by AugustinMauroy** [reply]
I mean by politic, is the policy about third dep on biome

**Reply by dyc3** [reply]
We generally try to minimize our dependency tree, but we don't explicitly prohibit bringing in new dependencies. It has to provide enough value for the project as a whole. For example, we won't bring in a dependency to be used for a single lint rule, but we would bring in a dependency for a specific data structure that someone else has already implemented.

All of biome's language parsers are implemented by us currently. If we were to bring in a package for parsing a new language, it would have to fit in to the rest of biome's parsing infrastructure. Like @ematipico said, PRs welcome.

**Reply by SalahAdDin** [reply]
No plans then.

**Comment by AugustinMauroy** [comment]
And what about "metadata" ?

**Reply by dyc3** [reply]
Are you talking about frontmatter?

**Reply by AugustinMauroy** [reply]
Yeah yeah

**Reply by dyc3** [reply]
IIRC that's a part of the common mark spec so yeah, support for that is planned. Potentially also requires #3334 and maybe even #2365
